cheatsheets/Forgot_Password_Cheat_Sheet.md

```diff
@@ -1,60 +1,137 @@
 
-This introduces an *out-of-band* communication channel and adds defense-in-depth as it is another barrier for a hacker to overcome. If the bad guy has somehow managed to successfully get past steps 1 and 2, he is unlikely to have compromised the side-channel. It is also a good idea to have the random code which your system generates to only have a limited validity period, say no more than 20 minutes or so. That way if the user doesn't get around to checking their email and their email account is later compromised, the random token used to reset the password would no longer be valid if the user never reset their password and the *reset password* token was discovered by an attacker. 
+- The service should redirect the user after using that token and replacing it with a cookie in order to avoid [Cross-domain Referer leakage](https://portswigger.net/kb/issues/00500400_cross-domain-referer-leakage). This [attack](https://hackerone.com/reports/209352) occurs when private information is sent in the querystring, and third party services are called, thus *leaking* the querystring in the `Referer` header.
 
-The user should always be sent a token rather than a new password for their account. There are numerous problems with directly sending the user a password, especially via email. These include:
+### User Resets Password
 
-- The transport mechanism used to send the password (email, SMS, etc) may not be encrypted.
-- The password may be stored in the users' mailbox or device for an indefinite period.
-- The password may be accessible to IT staff or other users who have access to the mailbox or device.
-- The password may be logged by third parties such as mail security scanning platforms.
+Once the user is validated through the provided token or code, they should reset their password to a new secure one. In order to secure this last step, the measures should be taken:
 
-Of course, by all means, once a user's password has been reset, the randomly-generated token should no longer be valid.
+- Validate that a secure password policy is in place.
+- The user should confirm the password they set by writing it twice.
+- Update and store the password following [secure practices](Password_Storage_Cheat_Sheet.md).
+- Send the user an email informing them that their password has been reset.
+- Ask the user to re-login. Don't auto-login users on password reset!
+- Ask the user if they want to invalidate all of the sessions, or invalidate the sessions by default without prompting the user.
 
-## Step 4) Allow user to change password in the existing session
+## Methods
```

Comment:
This is one the same level as "Methods Implementation" and "Tokens and Codes Secure Practices", which doesn't really make sense. This part of the CS doesn't flow very well.